r2-core/src/main/java/com/linkedin/r2/filter/transport/ServerRetryFilter.java

```diff
@@ -75,13 +116,160 @@ public void onStreamError(Throwable ex,
+   * a ratio of how many requests are being retried in the cluster. The ratio is then compared with
+   * {@link ServerRetryTracker#_maxRequestRetryRatio} to make a decision on whether or not to retry in the
+   * next interval. When calculating the ratio, it looks at the last {@link ServerRetryTracker#_aggregatedIntervalNum}
+   * intervals by aggregating the recorded requests.
+   */
+  private static class ServerRetryTracker
+  {
+    private final int _retryLimit;
+    private final int _aggregatedIntervalNum;
+    private final double _maxRequestRetryRatio;
+    private final long _updateIntervalMs;
+    private final Clock _clock;
+
+    private final Object _counterLock = new Object();
+    private final Object _updateLock = new Object();
+
+    @GuardedBy("_updateLock")
+    private volatile long _lastRollOverTime;
+    private boolean _isBelowRetryRatio;
+
+    @GuardedBy("_counterLock")
+    private final LinkedList<int[]> _retryAttemptsCounter;
+    private final int[] _aggregatedRetryAttemptsCounter;
+
+    private ServerRetryTracker(int retryLimit, int aggregatedIntervalNum, double maxRequestRetryRatio, long updateIntervalMs, Clock clock)
+    {
+      _retryLimit = retryLimit;
+      _aggregatedIntervalNum = aggregatedIntervalNum;
+      _maxRequestRetryRatio = maxRequestRetryRatio;
+      _updateIntervalMs = updateIntervalMs;
+      _clock = clock;
+
+      _aggregatedRetryAttemptsCounter = new int[_retryLimit + 1];
+      _retryAttemptsCounter = new LinkedList<>();
+      _retryAttemptsCounter.add(new int[_retryLimit + 1]);
+    }
+
+    public void add(int numberOfRetryAttempts)
+    {
+      if (numberOfRetryAttempts <= _retryLimit)
+      {
+        synchronized (_counterLock)
+        {
+          _retryAttemptsCounter.getLast()[numberOfRetryAttempts] += 1;
+        }
+      } else
+      {
+        LOG.warn("Unexpected number of retry attempts: " + numberOfRetryAttempts + ", current retry limit: " + _retryLimit);
+      }
+
+      updateRetryDecision();
+    }
+
+    public void rollOverStats()
+    {
+      // rollover the current interval to the aggregated counter
+      synchronized (_counterLock)
+      {
+        int[] intervalToAggregate = _retryAttemptsCounter.getLast();
+        for (int i = 0; i < _retryLimit; i++)
+        {
+          _aggregatedRetryAttemptsCounter[i] += intervalToAggregate[i];
+        }
+
+        if (_retryAttemptsCounter.size() > _aggregatedIntervalNum)
+        {
+          // discard the oldest interval
+          int[] intervalToDiscard = _retryAttemptsCounter.removeFirst();
+          for (int i = 0; i < _retryLimit; i++)
+          {
+            _aggregatedRetryAttemptsCounter[i] -= intervalToDiscard[i];
+          }
+        }
+
+        // append a new interval
+        _retryAttemptsCounter.addLast(new int[_retryLimit + 1]);
+      }
+    }
+
+    public boolean isBelowRetryRatio()
+    {
+      updateRetryDecision();
```

Comment:
This is necessary because if there aren't many requests coming in for a period of time, the information in the tracker is stale. The next time we make the retry decision, we have to first rollover the stale windows.

Reply:
Right, but that staleness would only apply if it was an older client not sending the header right? So the calculation at the end of the request can be skipped on requests when it's performed at the beginning.

Reply:
Right. Removed `updateRetryDecision` here